### PR TITLE
feat(contacts): move contacts to scim from dss

### PR DESCRIPTION
**Pull request by sreenara**
# COMPLETES #[SPARK-560428](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-560428)

## This pull request addresses

The contact client within the SDK using the @webex/internal-plugin-dss to perform resolutions on organization contacts. The problem here is that the DSS has always been internal and does not expose any public APIs. 
According to Janelle, who is the PM for DSS.
> It is not new news that DSS is a. private api and always has been.  We do offer both SCIM and Contacts apis Imagicle can call which is in fact what DSS is doing.  This DSS apis were not designed or created to be public apis and to make them public would need a full architectural review and plan to do so. 

Due to this, the advice for the SDK is to move to using SCIM, which is publicly available in the [SCIM API documentation](https://developer.webex.com/docs/scim-2-overview) on developer portal.

## by making the following changes

1. Re-use the `scimQuery()` function in Utils to do the contact resolution
2. New types created for the SCIM Response
3. Remove reliance on internal-plugin-dss and all function calls relating to this
4. Mock webex object and test changes for the above

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested
1. Get contacts - cloud and custom. Checked the details of the retrieved contacts
2. Create a custom contact
3. Create a cloud contact
4. Create a contact group

Vidcast link: https://app.vidcast.io/share/655b203c-5c2c-428c-9b8b-cb3636865df3

Additional tests:
1. Outgoing call to PSTN was successful. Outgoing call to internal destination successful.
2. Incoming call from internal extension successful
3. Hold/Resume and 2 types of transfer successful
4. Mute, DTMF, Caller ID resolution successful

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: docs/samples/calling/app.js [hunk 1]
```diff
@@ -1301,6 +1301,10 @@ async function createCustomContact() {
       type: 'work',
       value: formData.get('phone')
     }],
+    emails: [{
```
Comment (sreenara) [inline docs/samples/calling/app.js:1304]: We weren't adding emails to a custom contact even though the samples page has a field for it. Fixed this issue.

File: packages/calling/src/CallingClient/calling/CallerId/types.ts [hunk 2]
```diff
@@ -18,31 +18,6 @@ export type PhotoType = {
   value: string;
 };
 
-export type ResourceType = {
```
Comment (sreenara) [inline packages/calling/src/CallingClient/calling/CallerId/types.ts:21]: Moved this to common types as it's being used by both calling client and contacts now.

File: packages/calling/src/Contacts/ContactsClient.ts [hunk 3]
```diff
@@ -253,75 +260,59 @@ export class ContactsClient implements IContacts {
...
-        emails,
-        encryptionKeyUrl,
-        firstName,
-        groups,
-        lastName,
-        manager,
-        ownerId,
-        phoneNumbers,
-        sipAddresses,
-        title: jobTitle,
```
Comment (sreenara) [inline packages/calling/src/Contacts/ContactsClient.ts:316]: title is not available in SCIM response. So I've removed this.

File: packages/calling/src/SDKConnector/types.ts [hunk 4]
```diff
@@ -82,10 +75,6 @@ export interface WebexSDK {
         };
       };
     };
-    dss: {
```
Comment (sreenara) [inline packages/calling/src/SDKConnector/types.ts:85]: we shouldn't be using dss as it is internal

File: packages/webex/src/calling.js [hunk 5]
```diff
@@ -5,7 +5,6 @@ import EventEmitter from 'events';
 require('@webex/internal-plugin-device');
 require('@webex/internal-plugin-mercury');
 require('@webex/internal-plugin-encryption');
-require('@webex/internal-plugin-dss');
```
Comment (sreenara) [inline packages/webex/src/calling.js:8]: calling shouldn't use dss

File: packages/calling/src/common/types.ts [hunk 6]
```diff
@@ -30,14 +32,13 @@ export enum CALLING_BACKEND {
...
 export type CallId = string; // guid;
 export type CorrelationId = string;
-export type SipAddress = string;
 export enum CallType {
   URI = 'uri',
   TEL = 'tel',
 }
 export type CallDetails = {
   type: CallType;
-  address: SipAddress; // sip address
```
Comment (sreenara) [inline packages/calling/src/common/types.ts:40]: This is because we now have a SipAddress type.

File: packages/calling/src/common/types.ts [hunk 7]
```diff
@@ -30,14 +32,13 @@ export enum CALLING_BACKEND {
 export type DeviceList = unknown;
 export type CallId = string; // guid;
 export type CorrelationId = string;
-export type SipAddress = string;
```
Comment (sreenara) [inline packages/calling/src/common/types.ts:33]: unnecessary type, so removed it.

File: packages/calling/src/Contacts/ContactsClient.ts [hunk 8]
```diff
@@ -253,75 +260,59 @@ export class ContactsClient implements IContacts {
...
+        const {sipAddresses} = filteredContact[SCIM_WEBEXIDENTITY_USER];
+        const firstName = filteredContact.name.givenName;
+        const lastName = filteredContact.name.familyName;
+        const manager = filteredContact[SCIM_ENTERPRISE_USER].manager.displayName;
+        const department = filteredContact[SCIM_ENTERPRISE_USER].department;
+
+        let avatarURL = '';
+        if (photos?.length) {
+          avatarURL = photos[0].value;
+        }
```
Comment (Shreyas281299) [inline packages/calling/src/Contacts/ContactsClient.ts:284]: Nitpick
```suggestion
        let avatarURL = photos?.length ? photos?.length : ""
```

File: packages/calling/src/Contacts/ContactsClient.ts [hunk 9]
```diff
@@ -253,75 +260,59 @@ export class ContactsClient implements IContacts {
...
-      };
-
-      return cloudContact;
     } catch (error: any) {
       Logger.error(new ExtendedError(error.message, {}, ERROR_TYPE.DEFAULT), {});
 
       return null;
     }
+
+    return finalContactList;
```
Comment (Shreyas281299) [inline packages/calling/src/Contacts/ContactsClient.ts:315]: Would it be better to put this in a .finally block

**Review by Shreyas281299 — APPROVED** [review]

**Review by Kesari3008 — APPROVED** [review]

File: packages/calling/src/Contacts/ContactsClient.ts (hunk 9 above)
Comment (sreenara) [reply packages/calling/src/Contacts/ContactsClient.ts:315]: @Shreyas281299 are there any advantages to doing this?

**Pull request merged**